node/src/signing/queue.rs

```diff
@@ -0,0 +1,819 @@
+}
+
+/// Progress that persists across attempts.
+#[derive(Default)]
+pub struct SignatureComputationProgress {
+    /// Number of attempts that have been made to generate the signature.
+    /// This is used to abort after too many attempts.
+    pub attempts: u64,
+    /// The computed response, if any. This is used to prevent unnecessary recomputation,
+    /// if all that's needed is to submit the response to chain.
+    pub computed_response: Option<ChainRespondArgs>,
+    /// The time and when the last response was submitted to chain.
+    /// This is used to delay the next retry as well as debugging.
+    pub last_response_submission: Option<near_time::Instant>,
+}
+
+impl QueuedSignatureRequest {
+    pub fn new(
+        clock: &near_time::Clock,
+        request: SignatureRequest,
+        block_hash: CryptoHash,
+        block_height: u64,
+        all_participants: &[ParticipantId],
+    ) -> Self {
+        let leader_selection_order = Self::leader_selection_order(all_participants, request.id);
+        tracing::debug!(target: "signing", "Leader selection order for request {:?} from block {}: {:?}", request.id, block_height, leader_selection_order);
+
+        Self {
+            request,
+            block_hash,
+            block_height,
+            leader_selection_order,
+            computation_progress: Arc::new(Mutex::new(SignatureComputationProgress::default())),
+            next_check_due: clock.now(),
+            active_attempt: Weak::new(),
+        }
+    }
+
+    /// Computes the leader selection order for a given signature request.
+    /// This will be a different pseudorandom order for each signature request.
+    fn leader_selection_order(
+        participants: &[ParticipantId],
+        signature_request_id: CryptoHash,
+    ) -> Vec<ParticipantId> {
+        let mut leader_selection_hashes = participants
+            .iter()
+            .map(|p| (Self::leader_selection_hash(p, signature_request_id), *p))
+            .collect::<Vec<_>>();
+        leader_selection_hashes.sort();
+        leader_selection_hashes
+            .into_iter()
+            .map(|(_, p)| p)
+            .collect()
+    }
+
+    fn leader_selection_hash(
+        participant_id: &ParticipantId,
+        signature_request_id: CryptoHash,
+    ) -> u64 {
+        let mut h = Sha256::new();
+        h.update(participant_id.raw().to_le_bytes());
+        h.update(signature_request_id.0);
+        let hash: [u8; 32] = h.finalize().into();
+        u64::from_le_bytes(hash[0..8].try_into().unwrap())
+    }
+
+    /// Selects the leader given the current state of the network.
+    fn current_leader(&self, eligible_leaders: &HashSet<ParticipantId>) -> Option<ParticipantId> {
+        for candidate_leader in &self.leader_selection_order {
+            if eligible_leaders.contains(candidate_leader) {
+                return Some(*candidate_leader);
+            }
+        }
+        None
```

Comment:
What happens when the current leader is `None`? Does the request just fail?

Reply:
If the current leader is None, the request is ignored for now, but remains in the queue.